controllers/cluster_controller_phases.go

```diff
@@ -74,6 +74,12 @@ func (r *ClusterReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		return nil, err
 	}
 
+	// if external ref is paused, return error.
+	if util.IsPaused(cluster, obj) {
+		logger.V(3).Info("Infrastructure ref is paused")
+		return nil, errors.New("Infrastructure ref is paused")
```

Comment:
We can probably just return without an error

Reply:
we don't need to requeue the reconciliation? and in that case just return `nil, nil`?

Reply:
+1 to not returning an error. We definitely want to log, but we should get a reconciliation request if the infrastructure resource or the Cluster is updated to remove the condition for the pause.

Reply:
if this func does not return an error, then the calling reconcile func on the will try to continue processing and give an error, 
for example in this case here: https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/cluster_controller_phases.go#L149, the infraConfig is nil with no error

what should we do in that case?

Reply:
Ah, I'm wondering if we need to add a new Error type that we can swallow at the caller then?

Reply:
Or return a `reconcileExternalResult` that contains the object, paused, and error.

Reply:
I can try doing that and see what you all think about it

Reply:
> Or return a reconcileExternalResult that contains the object, paused, and error.

Would we want to lump the error into a return type like that? That would introduce different error handling than we currently have elsewhere.

Reply:
We could leave the error outside the struct

Reply:
I'd be hesitant to return (unstructured, bool, error) though - the anonymous nature of the bool makes it hard to know what it's for.

Reply:
It should be fine to keep finishing up the reconciliation loop, but avoid to read/write to the paused object

Reply:
if we don't return an error from the `reconcileExternal` func, we can't make it avoid the following steps after the `get`. I suggest returning a custom type of error and then in the calling func swallowing that and not doing a requeue as Jason mentioned above

Reply:
IMHO struct with unstructured + bool >>>>> custom error

Reply:
ok, will try that instead

Reply:
Given that this code needs to be revisited, because it's repeated in multiple files, I'd suggest the following:

- Create a new type in `external` package called
```golang
type ReconcileExternalOutput struct {
 // add Error, Result, and Paused
}
```
- Change reconcileExternal methods in clsuter_controller_phases.go and machine_controller_phases.go to use the type above

Reply:
At some point in the future, we'll need to figure out how to refactor this code, and this is probably a good step in that direction